### PR TITLE
Lerignoux/pij 1/pijon json migration

### PR DESCRIPTION
**Pull request by lerignoux**

### CODE REVIEW

File: pijon/pijon.py [hunk 1]
```diff
@@ -1,2 +1,82 @@
...
+import sys
+
+
+log = logging.getLogger(__name__)
+
+
 class Pijon():
-    pass
+
+    def __init__(self, migration_folder, files_format='migration_\d+_.*'):
```
Comment (gdraynz) [inline pijon/pijon.py]: I think imposing a file format would be better, let Pijon manage everything migration-related.
example for `0001_something.py`
```python
FILE_FORMAT = r'(?P<ident>\d{4})_(?P<name>\w+)\.py'
```

File: pijon/pijon.py [hunk 2]
```diff
@@ -1,2 +1,82 @@
...
+        self.folder = migration_folder
+        self.files_format = files_format
+        self.migrations = self.list_migrations()
+
+    @classmethod
+    def list(cls, migrations_folder='./pijon', file_format='migration_\d+_.*'):
+        return Pijon(migrations_folder).migrations
+
+    def list_migrations(self):
+        files = os.listdir(self.folder)
```
Comment (gdraynz) [inline pijon/pijon.py]: Use sorted(os.listdir(self.folder)

File: pijon/pijon.py [hunk 3]
```diff
@@ -1,2 +1,82 @@
...
+        return migrations
+
+    def migrate_step(self, step, data):
+        """
+        Perform a migration step
+        """
+        log.debug("Migrating step {}".format(step))
+        folder = re.findall('[^\/]+$', self.folder)[0]
+        module = "{folder}.{step}".format(folder=folder, step=step)
+        importlib.import_module(module)
```
Comment (gdraynz) [inline pijon/pijon.py]: You might want to use `importlib.machinery`
```python
def load_module(module_name, path):
    return machinery.SourceFileLoader(module_name, path).load_module(module_name)
mod = load_module('my_module_name', 'migrations/')
mod.migrate(data)
```
`importlib.import_module`entirely depends on your current path, but we want to use any migration from any folder.
(in our case: nyuki are not libraries)

File: README.md [hunk 4]
```diff
@@ -1,2 +1,28 @@
 # pijon
 A json migration tool
+
```
Comment (gdraynz) [inline README.md]: Not fond of calling for a classmethod as a static method. Throw this method out of the class and use it as a helper:
```python
def migrate(data, folder):
  pijon = Pijon(...)
  pijon.migrate(...)
```
```python
import pijon
pijon.migrate(my_data, my_folder)
```

File: pijon/pijon.py [hunk 5]
```diff
@@ -1,2 +1,82 @@
...
+        perform all migrations registered
+        """
+        data = input_data
+        for step in self.migrations:
+            data = self.migrate_step(step, data)
+        return data
+
+    @classmethod
+    def migrate(cls, in_data, direct_input=False, direct_output=False, output_filename=None,
+                migrations_folder='./pijon', file_format='migration_\d+_.*'):
```
Comment (gdraynz) [inline pijon/pijon.py]: Doing `isinstance` check on data to migrate could also be more friendly, automatically resolve the data received and apply the proper migrations if its a dictionary, a filename, a file object...

File: README.md (hunk 4 above)
Comment (thavel) [inline README.md]: :+1:

File: pijon/pijon.py (hunk 1 above)
Comment (thavel) [inline pijon/pijon.py]: I don't agree with @gdraynz, I like the idea to be able to alter the default file format. Although, I think @gdraynz's default `FILE_FORMAT` regex is better. `migration_` is quite useless imo.

File: pijon/pijon.py (hunk 3 above)
Comment (thavel) [inline pijon/pijon.py]: Did know `importlib.machinery`, but it sounds like a better solution, indeed. :+1:

File: pijon/pijon.py (hunk 5 above)
Comment (thavel) [inline pijon/pijon.py]: :+1:

File: pijon/pijon.py [hunk 6]
```diff
@@ -1,2 +1,82 @@
...
-    pass
+
+    def __init__(self, migration_folder, files_format='migration_\d+_.*'):
+        sys.path.append(migration_folder)
+        self.folder = migration_folder
+        self.files_format = files_format
+        self.migrations = self.list_migrations()
+
+    @classmethod
+    def list(cls, migrations_folder='./pijon', file_format='migration_\d+_.*'):
```
Comment (thavel) [inline pijon/pijon.py]: @gdraynz comment could apply here to (use a module method to wrap this one).
Regarding the file format, better use a global variable to set this only once.

File: pijon/pijon.py [hunk 7]
```diff
@@ -1,2 +1,82 @@
...
+        else:
+            with open(in_data, 'r') as input_file:
+                input_data = json.loads(input_file.read())
+
+        result = pijon._migrate(input_data)
+        if direct_output:
+            return json.dumps(result)
+        else:
+            if output_filename is None:
+                output_filename = input_filename
```
Comment (thavel) [inline pijon/pijon.py]: `input_filename` isn't defined anywhere.

File: pijon/migration_template.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+def migrate(old_dict):
+    """
+    Put your migration step here.
+    the old_data is in old_dict in a dictionary form.
+    the new_data returned is the data after migrating both schema and data
+    """
+
+    # Put the migration between old_data and new_data here
+
+    return news_dict
```
Comment (thavel) [inline pijon/migration_template.py]: Typo

File: pijon/migration_template.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+def migrate(old_dict):
+    """
+    Put your migration step here.
+    the old_data is in old_dict in a dictionary form.
+    the new_data returned is the data after migrating both schema and data
+    """
+
+    # Put the migration between old_data and new_data here
+
```
Comment (thavel) [inline pijon/migration_template.py]: Better set `new_dict = old_dict` to have a valid Python file, even if it won't be used, don't you think?

**Comment by thavel** [comment]
There is no "from where to start migrations" mechanism.
Would be nice to have the targeted JSON schema included in the migration template:
```python
target_schema = { ... }

def migrate(in_data):
    # Script content
```

Therefore, we could detect the next valid migration to apply by checking schemas one after another from the first one to be last. And then, apply only required migration.

Otherwise, the current data could be in a state that wouldn't be compatible with and old migration script.

File: pijon/pijon.py (hunk 1 above)
Comment (gdraynz) [inline pijon/pijon.py]: Then we just won't have any idea of what we are reading. Is there a unique identifier that allow us to migrate in the right order ? what if these files are not python files I can import ? etc.

Comment (thavel) [inline pijon/pijon.py]: We should allow custom order (and therefore custom file format). But I agree, we should move the file extension out of the regex to force only python files.

**Pull request merged**